### DISCUSSION TITLE
Operator / word tokens

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#11] [General]
# Operator / word tokens

Is there any legitimate reason why qpdf treats tokens such as `m25-724.78689.35` or `8�g��_�W~0�` (38 a1 05 67 91 f5 5f f8 57 7e 30 ff) as word tokens / operator objects? 

According to 7.8.2 of the spec, 

> Ordinarily, when a conforming reader encounters an operator in a content stream that it does not recognize, an
error shall occur. A pair of compatibility operators, BX and EX (PDF 1.1), shall modify this behaviour .

This would suggest to me that such tokens should be treated as error tokens rather than as word tokens. While qpdf may not want to check every word token for validity, the vast majority of invalid tokens are easily detected, which in turn would make it much easier to detect corrupt streams and avoid lengthy pointless processing,

I would suggest something like

```mermaid
stateDiagram-v2
  [*] --> word : a-zA-Z
  [*] --> word_final : `"
  word --> word : a-zA-Z
  word --> word_final : *
  word --> [*]: delim
  word_final --> [*] :delim
``` 
with all other transitions leading to tt_bad.

**Comment by jberkenbilt** [comment]
No deep reason. It was a long time before qpdf had the capability to parse content streams at all, and in the early days, I tried really hard to make sure I never further damaged already damaged content streams. But if the parser can get more intelligent and create "bad" tokens (or whatever we call them) and leave word tokens for syntactically valid word/operator tokens, I agree that would be an improvement, and it should make no difference at all for parsing valid PDF files or gracefully preserving junk encountered in content streams. We'd obviously have to document this as an incompatible change, but I'd classify this as a bug fix. ("qpdf no longer incorrectly classifies unrecognized strings of characters as word tokens when they are not considered valid operator tokens by the PDF spec" -- or something along those lines). I haven't studied the parser/tokenizer much lately. You've more or less rewritten them since I've last been in the code, so I don't have in my head what we do with noise.
